frontend/pages/TagView.js

```diff
@@ -1,14 +1,181 @@
 import React from "react";
-import ParisMap from "../components/ParisMap";
+import ParisMap, {MAPSQUARE_HEIGHT, MAPSQUARE_WIDTH} from "../components/ParisMap";
 import * as PropTypes from "prop-types";
 import PhotoViewer from "../components/PhotoViewer";
+import LoadingPage from "./LoadingPage";
+import {GeoJSON, Popup, Rectangle} from "react-leaflet";
+
+function densityOverlay(mapSquareData) {
+    const sortedMapData = Object.values(mapSquareData)
+    .sort((a, b) => a.num_photos - b.num_photos);
+    // Gets the max number of photos in a single square out of all squares to form buckets later
+    const maxNumOfPhotos = sortedMapData && sortedMapData.length
+        ? sortedMapData[sortedMapData.length - 1].num_photos
+        : 0;
+    // Creating 5 buckets based on lowest to highest number of photos per square
+    const twentyPctMax = Math.round(0.2 * maxNumOfPhotos);
+    const fortyPctMax = Math.round(0.4 * maxNumOfPhotos);
+    const sixtyPctMax = Math.round(0.6 * maxNumOfPhotos);
+    const eightyPctMax = Math.round(0.8 * maxNumOfPhotos);
+    return (<>
+        {
+            sortedMapData.map((mapSquareData) => {
+                const index = mapSquareData.number;
+                const coords = mapSquareData.topLeftCoords;
+                const numberOfPhotos = mapSquareData.num_photos;
+
+                const mapSquareBounds = [
+                    [coords.lat, coords.lng],
+                    [coords.lat - MAPSQUARE_HEIGHT, coords.lng - MAPSQUARE_WIDTH]
+                ];
+                const link = "/map_square/" + index;
+                let mapSquareBucket = "";
+                // set of conditionals to calculate photo density for heat map
+                if (numberOfPhotos > 0 && numberOfPhotos <= twentyPctMax) {
```

Comment:
These buckets can definitely be implemented automatically: instead of using multiple `if/else` blocks, you can use division to map numberOfPhotos to bucket #. This has a few benefits: (1) it's cleaner, and (2) is easier to change in the future i.e. if you want to change buckets from 5 to 10.